### PR TITLE
Automatically synthesize ilproj assembly names for tests

### PR DESCRIPTION
**Pull request by kg**
* Add `-ANAME:AssemblyName` switch to ilasm
* Add `SynthesizeIlasmAssemblyName` msbuild property to the IL SDK that overrides the assembly name of the assembled IL to match the name of the current project

Motivation:
We have a bunch of ilproj tests that have random `.assembly` directives in them with names that don't match the name of the actual .dll file. This creates a problem for the crossgen2 composite tests, because the actual name of the dll isn't encoded anywhere in the binary.
We also have a bunch of tests where we build the same .il file multiple times with different names, i.e. `varargsupport_r` and `varargsupport` both being built from `varargsupport.il`. In this scenario it's impossible for the `.assembly` directive to encode the correct name unless you jump through some elaborate hoops, which not all of our tests are doing. The result is that during the crossgen2 tests, `varargsupport.dll` will be loaded (because the assembly name encoded in the reference and binaries is `varargsupport`) but the actual file is `varargsupport_r.dll` so we'll fail.
To solve that, this PR adds an `-ANAME` switch which allows overriding the assembly name of the .il file(s) being compiled, ignoring any directives being used in them. And then it updates the SDK to allow setting this switch's value manually, or set it automatically (synthesize it) from the name of the current msbuild project.

Once the synthesize property is applied to all the relevant test projects it should fix a bunch of failures on the crossgen2 outerloop tests. I verified it locally for the one folder I've applied it to in this PR.

These changes (to ILAsm and the IL sdk) will need to flow before the tests actually start passing on outerloop, when I tested locally I changed global.json and eng/versions.props to `10.0.0-dev` for the relevant packages in order to test it manually. I don't think it's possible for the flow to happen until this PR lands.

### CODE REVIEW

**Comment by JulieLeeMSFT** [comment]
cc @amanasifkhalid.

**Comment by amanasifkhalid** [comment]
@kg is this ready for review, or would you prefer I wait for this to come out of draft mode?
(I'm unfamiliar with this code, so you may want to request a review from someone else as well, once you're ready.)

**Comment by kg** [comment]
> @kg is this ready for review, or would you prefer I wait for this to come out of draft mode? (I'm unfamiliar with this code, so you may want to request a review from someone else as well, once you're ready.)

It's not ready yet, thank you for looking. I'm working on fixing some crossgen outerloop test failures, which will require a bunch of changes to IL SDK based tests.

**Comment by kg** [comment]
cc @jkoritzinsky

**Comment by jkoritzinsky** [comment]
I like the idea here! Really improves maintenance of the IL tests.

Maybe we can do a pass over the IL tests and remove the .assembly directives afterwards and then merge many of the IL tests into the same ilproj (as ilasm can take multiple input files)?

**Comment by kg** [comment]
> I like the idea here! Really improves maintenance of the IL tests.
> 
> Maybe we can do a pass over the IL tests and remove the .assembly directives afterwards and then merge many of the IL tests into the same ilproj (as ilasm can take multiple input files)?

That's the goal, as a starting point I want to turn this on for all the tests that are failing, and then remove the weird tricks that we're currently using for this (some variation on `#define ASSEMBLY_NAME xxx` in a dummy il file)

**Comment by kg** [comment]
Just in case "ready for review" didn't send an email, this is waiting for review

**Review by jkoritzinsky — APPROVED** [review]

**Pull request merged**
